File: Solution/Javascript/746. Min Cost Climbing Stairs.js
/**
 * @param {number[]} cost
 * @return {number}
 */
// var minCostClimbingStairs = function (cost) {
//     var a = 0, b = 0, min = 0;
//     for (let c in cost) {
//         b = a;
//         a = cost[c] + min;
//         min = Math.min(a, b);
//     }
//     return min;
// };

var minCostClimbingStairs = function (cost) {
    let dp = [];
    dp.length = cost.length;
    dp.fill(0);
    dp[0] = cost[0];
    dp[1] = cost[1];
    for (let i = 2; i < cost.length; i++) {
        dp[i] = cost[i] + Math.min(dp[i - 1], dp[i - 2]);
    }
    return Math.min(dp[cost.length - 1], dp[cost.length - 2]);
};

//let cost = [10, 15, 20];
let cost = [1, 100, 1, 1, 1, 100, 1, 1, 100, 1];
console.log(minCostClimbingStairs(cost));

File: Solution/Python/746. Min Cost Climbing Stairs.py
# dp
class Solution:
    def minCostClimbingStairs(self, cost):
        """
        :type cost: List[int]
        :rtype: int
        """
        dp = [0] * len(cost)
        dp[0] = cost[0]
        dp[1] = cost[1]
        for i in range(2, len(cost)):
            dp[i] = cost[i] + min(dp[i - 1], dp[i - 2])
        return min(dp[-1], dp[-2])


s = Solution()
cost = [1, 100, 1, 1, 1, 100, 1, 1, 100, 1]
res = s.minCostClimbingStairs(cost)
print(res)
